helix-view/src/document.rs

```diff
@@ -111,6 +115,144 @@ impl FromStr for Mode {
+    let mut builder = RopeBuilder::new();
+
+    // By default, the encoding of the text is auto-detected via the
+    // `chardetng` crate which requires sample data from the reader.
+    // As a manual override to this auto-detection is possible, the
+    // same data is read into `buf` to ensure symmetry in the upcoming
+    // loop.
+    let (encoding, mut decoder, mut slice, mut is_empty) = {
+        let read = reader.read(&mut buf)?;
+        let is_empty = read == 0;
+        let encoding = encoding.unwrap_or_else(|| {
+            let mut encoding_detector = chardetng::EncodingDetector::new();
+            encoding_detector.feed(&buf, is_empty);
+            encoding_detector.guess(None, true)
+        });
+        let decoder = encoding.new_decoder();
+
+        // If the amount of bytes read from the reader is less than
+        // `buf.len()`, it is undesirable to read the bytes afterwards.
+        let slice = &buf[..read];
+        (encoding, decoder, slice, is_empty)
+    };
+
+    // `RopeBuilder::append()` expects a `&str`, so this is the "real"
+    // output buffer. When decoding, the number of bytes in the output
+    // buffer will often exceed the number of bytes in the input buffer.
+    // The `result` returned by `decode_to_str()` will state whether or
+    // not that happened. The contents of `buf_str` is appended to
+    // `builder` and it is reused for the next iteration of the decoding
+    // loop.
+    //
+    // As it is possible to read less than the buffer's maximum from `read()`
+    // even when the end of the reader has yet to be reached, the end of
+    // the reader is determined only when a `read()` call returns `0`.
+    let buf_str = unsafe { std::str::from_utf8_unchecked_mut(&mut buf_out[..]) };
+    let mut total_read = 0usize;
+    let mut total_written = 0usize;
+    loop {
+        let (result, read, written, ..) = decoder.decode_to_str(
+            &slice[total_read..],
+            &mut buf_str[total_written..],
+            is_empty,
+        );
+
+        // These additive quantaties are needed in case the `result` was
+        // `encoding_rs::CoderResult::OutputFull` as they act as "cursors"
+        // into the input and output buffers respectively. Otherwise, it
+        // would be trapped in an endless read loop or would overwrite
+        // the output buffer.
+        total_read += read;
+        total_written += written;
+        match result {
+            encoding_rs::CoderResult::InputEmpty => {
+                debug_assert_eq!(slice.len(), total_read);
+                let read = reader.read(&mut buf)?;
+                slice = &buf[..read];
+                is_empty = read == 0;
+                total_read = 0;
+            }
+            encoding_rs::CoderResult::OutputFull => {
+                debug_assert!(slice.len() > total_read);
+                builder.append(&buf_str[..total_written]);
+                total_written = 0;
+            }
+        }
```

Comment:
I believe you still need the second nested loop here, like you had in the previous version I looked at.  Specifically, you need to keep calling `decode_to_str()` untl you get `InputEmpty`, to ensure that you've processed the entire input buffer before reading more data into it from disk.

Reply:
Ah, I didn't notice this comment. I think it should be fine because it will continuous call `decode_to_str()` until `slice` has been fully decoded, upon which it'll read from the reader and check whether the end of stream was reached by checking if it the outputted value equates to 0.

Reply:
Oh!  Got it.  Okay, yeah, I think this is correct then.  Somehow I missed that you had moved the `read()` call into the `InputEmpty` case.

Personally, I find this version of the code harder to follow than what you had before.  Your previous version made the separate steps really clear by having them separated out in the code as well:

1. Read data from disk.
2. Fully process that data.
3. Return to step 1 until done.

This version weaves the steps together in the code and IMO makes it far less clear what's going on.  I don't think getting rid of nested loops is intrinsically a good thing--it's only good when it helps either readability or efficiency.  And this case neither is the case, as far as I can tell.  I'd recommend going back to the nested loop construction.  But I won't block merging on it.

Reply:
Btw, I really like that you separated these out into stand-alone functions that take `Read`/`Write` types.  I think that's a big win for making things easier to follow.

Reply:
Gotcha, I'll revert it. Good thing I made the documentation fairly agnostic of the actual code, I'd hate to have to rewrite it.

Reply:
Also marking this as resolved.  I misunderstood the code.